nsls2forge_utils/all_feedstocks.py

```diff
@@ -0,0 +1,109 @@
+'''
+This code is a rework of
+https://github.com/regro/cf-scripts/blob/master/conda_forge_tick/all_feedstocks.py
+for use by nsls-ii-forge's own auto-tick bot.
+'''
+import github3
+import os
+import datetime
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+def get_all_feedstocks_from_github(organization, username='', password=''):
+    '''
+    Gets all public feedstock repository names from the GitHub organization
+    (e.g. nsls-ii-forge).
+
+    Parameters
+    ----------
+    organization: str
+        Name of organization on GitHub.
+    username: str, optional
+        Name of user on GitHub for authentication.
+        Uses environment variables if not specified.
+    password: str, optional
+        Password of user on GitHub for authentication.
+        Uses environment variables if not specified.
+
+    Returns
+    -------
+    names: list
+        List of repository names that end with '-feedstock' (stripped).
+    '''
+    if username == '':
+        gh = github3.login(os.environ['USERNAME'], os.environ['PASSWORD'])
+    else:
+        gh = github3.login(username, password)
+    org = gh.organization(organization)
+    repos = org.repositories()
+    names = []
+    try:
+        for repo in repos:
+            name = repo.name
+            if name.endswith("-feedstock"):
+                name = name.split("-feedstock")[0]
+                logger.info(name)
+                names.append(name)
+    except github3.GitHubError:
+        msg = ["Github rate limited. "]
+        c = gh.rate_limit()["resources"]["core"]
+        if c["remaining"] == 0:
+            ts = c["reset"]
+            msg.append("API timeout, API returns at")
+            msg.append(
+                datetime.datetime.utcfromtimestamp(ts).strftime("%Y-%m-%dT%H:%M:%SZ"),
+            )
+        logger.warning(" ".join(msg))
+        raise
+    return names
+
+
+def get_all_feedstocks(organization, username='', password='', cached=False):
+    '''
+    Gets all feedstocks either from GitHub or from names.txt if flag is specified.
+
+    Parameters
+    ----------
+    organization: str
+        Name of organization on GitHub.
+    username: str, optional
+        Name of user on GitHub for authentication.
+        Uses environment variable if not specified.
+    password: str, optional
+        Password of user on GitHub for authentication.
+        Uses environment variable if username not specified.
```

Comment:
These two vars (`username` and `password`) can be "hidden" from the inputs by using `**kwargs`, which can be passed to `get_all_feedstocks_from_github` function below. Probably the organization can be added there as it's not used anywhere else in this function.

Reply:
Why should these be hidden if they are necessary to authenticate? I would have to check if they exist in kwargs and then check if they are None before attempting the login. The other way I only have to check if they are None before trying the login. I've never used kwargs before so I am unsure if this is how it works.

It would be like this if I used kwargs:
```
def get_all_feedstocks_from_github(organization, **kwargs):
    if 'username' not in kwargs:
        username = os.environ['USERNAME']  # or load from .netrc
    elif kwargs['username'] is None:
        username = os.environ['USERNAME']  # or load from .netrc
    else:
        username = kwargs['username']
. . .
    gh = github3.login(username, password)
. . .
```
vs. if I did it like this:
```
def get_all_feedstocks_from_github(organization, username=None, password=None):
    if username is None:
        username = os.environ['USERNAME']  # or load from .netrc
. . .
    gh = github3.login(username, password)
. . .
```

Reply:
I am talking about using `kwargs` in `get_all_feedstocks`, and pass them to `get_all_feedstocks_from_github` in a call, but the latter one will have all args/kwargs spelled explicitly.

In general, it's a matter of preference, but if you only need to pass the variables, it's a convenient Python language feature, which is extensively used by the Bluesky software stack as well as in scientific Python libraries. See e.g. https://github.com/dask/distributed/blob/44b2358e33a0738c4c70ca96db4242636245e07d/distributed/client.py#L500-L599. On the other hand, it will be less documented, and the users will have to go to the code, which actually uses it. I think in this specific case it's OK to use less vars in the wrapper function `get_all_feedstocks`, and just pass whatever is left via unpacking the `kwargs` (with `**kwargs`).

Reply:
Ah ok, that makes sense now. Will update. Thank you.

Reply:
I think it was too much information at the end of the day :)
The latest commit 335bb35 implements it differently than I was thinking. It'd be better to have the following signature in the `get_all_feedstocks`:
```py
def get_all_feedstocks(cached=False, **kwargs):
    ...
    names = get_all_feedstocks_from_github(**kwargs)
```
and the `get_all_feedstocks_from_github` will have the signature you originally had, where all kwargs are spelled out:
```py
get_all_feedstocks_from_github(organization, username=None, password=None):
    ...
```
Sorry for the confusion. Does it make sense now?

Reply:
I think the kwargs in general should be explicitly spelled in the signature of a function at their final destination, in this case it's the definition of `get_all_feedstocks_from_github`, as the kwargs are not used anyhow than that.

Reply:
I understand now, thank you.